src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java

```diff
@@ -465,33 +465,65 @@ private boolean repairSegment(final int rangeIndex, final UUID segmentId, Segmen
     LOG.debug("preparing to repair segment {} on run with id {}", segmentId, repairRunId);
 
     List<String> potentialCoordinators;
+    Set<String> segmentReplicas = segment.getReplicas().keySet();
     if (!repairUnit.getIncrementalRepair()) {
       // full repair
       try {
         potentialCoordinators = filterPotentialCoordinatorsByDatacenters(
                 repairUnit.getDatacenters(),
-                clusterFacade.tokenRangeToEndpoint(cluster, keyspace, segment));
+                clusterFacade.tokenRangeToEndpoint(cluster, keyspace, segment.getTokenRange()));
+
       } catch (RuntimeException e) {
         LOG.warn("Couldn't get token ranges from coordinator: #{}", e);
         return true;
       }
       if (potentialCoordinators.isEmpty()) {
         LOG.warn(
-            "Segment #{} is faulty, no potential coordinators for range: {}",
+            "Segment #{} is faulty, no potential coordinators for range: [{}, {}]",
             segmentId,
-            segment.toString());
+            segment.getStartToken(), segment.getEndToken());
         // This segment has a faulty token range. Abort the entire repair run.
         synchronized (this) {
           context.storage.updateRepairRun(
               context.storage.getRepairRun(repairRunId).get()
                   .with()
                   .runState(RepairRun.RunState.ERROR)
-                  .lastEvent(String.format("No coordinators for range %s", segment))
+                  .lastEvent(
+                      String.format(
+                          "No coordinators for range %s - %s",
+                          segment.getStartToken(),
+                          segment.getEndToken()))
                   .endTime(DateTime.now())
                   .build(repairRunId));
 
           killAndCleanupRunner();
         }
+
+        return false;
+      } else if (
+          !Sets.difference(
+            segmentReplicas,
+            potentialCoordinators.stream().collect(Collectors.toSet())).isEmpty()) {
+        LOG.warn(
+            "Segment #{} is faulty, replica set changed since repair was started: [{}, {}]",
+            segmentId,
+            segment.getStartToken(), segment.getEndToken());
+        // This segment has a faulty token range. Abort the entire repair run.
+        synchronized (this) {
+          context.storage.updateRepairRun(
+              context.storage.getRepairRun(repairRunId).get()
+                  .with()
+                  .runState(RepairRun.RunState.ERROR)
+                  .lastEvent(String.format("Replica set changed for segment %s on range [%s, %s]",
+                      segmentId,
+                      segment.getStartToken(),
+                      segment.getEndToken()))
+                  .endTime(DateTime.now())
+                  .build(repairRunId));
```

Comment:
This kills/aborts the whole repair run correct?  

What about other segments in it currently running? 

What about other repair runs registered but not yet started?

Reply:
yes, the whole repair will fail.
Running segments won't be impacted by this.

Other repair runs registered but not started will fail as well as soon as they hit a segment whose replica set is no longer valid.